File: python/src/ancilis/baselines/manager.py
from __future__ import annotations

import json
import uuid
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING, Any

from ancilis.baselines.drift import DriftDetector, _compute_control_stats, _dominant_result, _pass_rate
from ancilis.baselines.models import Baseline, ControlSnapshot, DriftReport

if TYPE_CHECKING:
    from ancilis.config import ResolvedConfig
    from ancilis.evidence.store import EvidenceStore

CREATE_BASELINES_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS baselines (
    baseline_id VARCHAR PRIMARY KEY,
    created_at VARCHAR NOT NULL,
    agent_id VARCHAR NOT NULL,
    overlay_id VARCHAR,
    label VARCHAR NOT NULL,
    control_snapshots JSON NOT NULL,
    metadata JSON,
    is_active BOOLEAN NOT NULL DEFAULT TRUE
);
"""


class BaselineManager:
    """Manages baseline snapshots and drift detection against the evidence store."""

    def __init__(self, evidence_store: EvidenceStore, config: ResolvedConfig) -> None:
        self._store = evidence_store
        self._config = config
        self._detector = DriftDetector()
        self._ensure_table()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _ensure_table(self) -> None:
        self._store._connection.execute(CREATE_BASELINES_TABLE_SQL)

    def _conn(self):  # type: ignore[return]
        return self._store._connection

    def _agent_id(self) -> str:
        return getattr(self._config, "agent_name", "") or "default"

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def create(
        self,
        label: str,
        overlay_id: str | None = None,
        evidence_window_hours: int = 168,
        metadata: dict[str, Any] | None = None,
    ) -> Baseline:
        """Snapshot current posture from the evidence window into a new baseline.

        Deactivates any existing active baseline for the same agent+overlay pair.
        """
        agent_id = self._agent_id()
        now = datetime.now(timezone.utc)
        window_start = (now - timedelta(hours=evidence_window_hours)).isoformat()
        window_end = now.isoformat()

        # Fetch evidence in the window
        conditions = ["timestamp >= ?"]
        params: list[Any] = [window_start]
        if overlay_id is not None:
            # Filter by overlay presence in active_overlays JSON array
            conditions.append("active_overlays LIKE ?")
            params.append(f"%{overlay_id}%")

        where = " WHERE " + " AND ".join(conditions)
        rows = self._conn().execute(
            "SELECT control_results FROM evidence_records" + where,
            params,
        ).fetchall()

        stats = _compute_control_stats(rows)

        snapshots = [
            ControlSnapshot(
                control_id=cid,
                result=_dominant_result(s),
                pass_rate=_pass_rate(s),
                total_evaluations=s["total"],
                evidence_window_start=window_start,
                evidence_window_end=window_end,
            )
            for cid, s in stats.items()
        ]

        # Deactivate existing active baseline for same agent+overlay
        if overlay_id is not None:
            self._conn().execute(
                "UPDATE baselines SET is_active = FALSE "
                "WHERE agent_id = ? AND overlay_id = ? AND is_active = TRUE",
                [agent_id, overlay_id],
            )
        else:
            self._conn().execute(
                "UPDATE baselines SET is_active = FALSE "
                "WHERE agent_id = ? AND overlay_id IS NULL AND is_active = TRUE",
                [agent_id],
            )

        baseline_id = str(uuid.uuid4())
        baseline = Baseline(
            baseline_id=baseline_id,
            created_at=now.isoformat(),
            agent_id=agent_id,
            overlay_id=overlay_id,
            label=label,
            control_snapshots=snapshots,
            metadata=metadata,
            is_active=True,
        )

        self._conn().execute(
            "INSERT INTO baselines "
            "(baseline_id, created_at, agent_id, overlay_id, label, "
            "control_snapshots, metadata, is_active) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            [
                baseline.baseline_id,
                baseline.created_at,
                baseline.agent_id,
                baseline.overlay_id,
                baseline.label,
                json.dumps([s.__dict__ for s in baseline.control_snapshots]),
                json.dumps(baseline.metadata) if baseline.metadata else None,
                baseline.is_active,
            ],
        )

        return baseline

    def list_baselines(self, overlay_id: str | None = None) -> list[Baseline]:
        """Return baselines for this agent, optionally filtered by overlay."""
        agent_id = self._agent_id()
        if overlay_id is not None:
            rows = self._conn().execute(
                "SELECT baseline_id, created_at, agent_id, overlay_id, label, "
                "control_snapshots, metadata, is_active FROM baselines "
                "WHERE agent_id = ? AND overlay_id = ? ORDER BY created_at DESC",
                [agent_id, overlay_id],
            ).fetchall()
        else:
            rows = self._conn().execute(
                "SELECT baseline_id, created_at, agent_id, overlay_id, label, "
                "control_snapshots, metadata, is_active FROM baselines "
                "WHERE agent_id = ? ORDER BY created_at DESC",
                [agent_id],
            ).fetchall()
        return [self._row_to_baseline(r) for r in rows]

    def get_baseline(self, baseline_id: str) -> Baseline:
        """Fetch a single baseline by ID. Raises KeyError if not found."""
        row = self._conn().execute(
            "SELECT baseline_id, created_at, agent_id, overlay_id, label, "
            "control_snapshots, metadata, is_active FROM baselines "
            "WHERE baseline_id = ?",
            [baseline_id],
        ).fetchone()
        if row is None:
            raise KeyError(f"Baseline not found: {baseline_id}")
        return self._row_to_baseline(row)

    def check_drift(
        self,
        baseline_id: str | None = None,
        overlay_id: str | None = None,
    ) -> DriftReport:
        """Run drift detection against the most recent (or specified) baseline.

        If baseline_id is None, uses the most recent active baseline for the
        agent (optionally scoped to overlay_id).
        """
        agent_id = self._agent_id()

        if baseline_id is not None:
            baseline = self.get_baseline(baseline_id)
        else:
            # Find most recent active baseline
            if overlay_id is not None:
                row = self._conn().execute(
                    "SELECT baseline_id, created_at, agent_id, overlay_id, label, "
                    "control_snapshots, metadata, is_active FROM baselines "
                    "WHERE agent_id = ? AND overlay_id = ? AND is_active = TRUE "
                    "ORDER BY created_at DESC LIMIT 1",
                    [agent_id, overlay_id],
                ).fetchone()
            else:
                row = self._conn().execute(
                    "SELECT baseline_id, created_at, agent_id, overlay_id, label, "
                    "control_snapshots, metadata, is_active FROM baselines "
                    "WHERE agent_id = ? AND is_active = TRUE "
                    "ORDER BY created_at DESC LIMIT 1",
                    [agent_id],
                ).fetchone()
            if row is None:
                raise KeyError("No active baseline found. Run 'ancilis baseline create' first.")
            baseline = self._row_to_baseline(row)

        # Fetch evidence since baseline was created
        since = baseline.created_at
        conditions = ["timestamp >= ?"]
        params: list[Any] = [since]
        if baseline.overlay_id is not None:
            conditions.append("active_overlays LIKE ?")
            params.append(f"%{baseline.overlay_id}%")

        where = " WHERE " + " AND ".join(conditions)
        rows = self._conn().execute(
            "SELECT control_results FROM evidence_records" + where,
            params,
        ).fetchall()

        current_stats = _compute_control_stats(rows)

        # Collect first failure timestamps per control
        first_failures: dict[str, str | None] = {}
        for cid in current_stats:
            fail_row = self._conn().execute(
                "SELECT MIN(timestamp) FROM evidence_records "
                "WHERE timestamp >= ? AND control_results LIKE ?",
                [since, f'%"control_id": "{cid}"%'],
            ).fetchone()
            first_failures[cid] = fail_row[0] if fail_row and fail_row[0] else None

        return self._detector.detect(baseline, current_stats, first_failures)

    def deactivate(self, baseline_id: str) -> None:
        """Mark a baseline as inactive."""
        self._conn().execute(
            "UPDATE baselines SET is_active = FALSE WHERE baseline_id = ?",
            [baseline_id],
        )

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    @staticmethod
    def _row_to_baseline(row: tuple[Any, ...]) -> Baseline:
        snapshots_raw = json.loads(row[5]) if isinstance(row[5], str) else row[5]
        snapshots = [ControlSnapshot(**s) for s in snapshots_raw]
        metadata_raw = row[6]
        metadata = json.loads(metadata_raw) if isinstance(metadata_raw, str) else metadata_raw
        return Baseline(
            baseline_id=row[0],
            created_at=row[1],
            agent_id=row[2],
            overlay_id=row[3],
            label=row[4],
            control_snapshots=snapshots,
            metadata=metadata,
            is_active=bool(row[7]),
        )

File: typescript/src/ancilis/baselines/manager.ts
/**
 * BaselineManager — snapshot and drift detection for evidence baselines.
 *
 * Trust model: baselines are operator assertions (policy intent), not a link in
 * the evidence hash chain. They are stored in the same DuckDB database as
 * evidence_records for convenience, but they carry no cryptographic guarantee.
 * Treat them as configuration, not as audit proof.
 */

import { randomUUID } from "node:crypto";
import type { EvidenceStore } from "../evidence/store.js";
import type { ResolvedConfig } from "../config/index.js";
import { computeControlStats, DriftDetector, dominantResult, passRate } from "./drift.js";
import type { Baseline, ControlSnapshot, DriftReport } from "./models.js";

const CREATE_BASELINES_TABLE_SQL = `
CREATE TABLE IF NOT EXISTS baselines (
    baseline_id VARCHAR PRIMARY KEY,
    created_at VARCHAR NOT NULL,
    agent_id VARCHAR NOT NULL,
    overlay_id VARCHAR,
    label VARCHAR NOT NULL,
    control_snapshots JSON NOT NULL,
    metadata JSON,
    is_active BOOLEAN NOT NULL DEFAULT TRUE,
    tenant_id VARCHAR
);`;

function rowToBaseline(row: Record<string, unknown>): Baseline {
  const snapshotsRaw = typeof row.control_snapshots === "string"
    ? JSON.parse(row.control_snapshots)
    : row.control_snapshots;
  const snapshots: ControlSnapshot[] = (snapshotsRaw as Array<Record<string, unknown>>).map(s => ({
    controlId: (s.control_id ?? s.controlId) as string,
    result: s.result as string,
    passRate: (s.pass_rate ?? s.passRate) as number,
    totalEvaluations: (s.total_evaluations ?? s.totalEvaluations) as number,
    evidenceWindowStart: (s.evidence_window_start ?? s.evidenceWindowStart) as string,
    evidenceWindowEnd: (s.evidence_window_end ?? s.evidenceWindowEnd) as string,
  }));
  const metaRaw = row.metadata;
  const metadata = metaRaw
    ? (typeof metaRaw === "string" ? JSON.parse(metaRaw) as Record<string, unknown> : metaRaw as Record<string, unknown>)
    : null;
  return {
    baselineId: row.baseline_id as string,
    createdAt: row.created_at as string,
    agentId: row.agent_id as string,
    overlayId: (row.overlay_id as string | null | undefined) ?? null,
    label: row.label as string,
    controlSnapshots: snapshots,
    metadata,
    isActive: Boolean(row.is_active),
  };
}

function snapshotsToJson(snapshots: ControlSnapshot[]): string {
  return JSON.stringify(snapshots.map(s => ({
    control_id: s.controlId,
    result: s.result,
    pass_rate: s.passRate,
    total_evaluations: s.totalEvaluations,
    evidence_window_start: s.evidenceWindowStart,
    evidence_window_end: s.evidenceWindowEnd,
  })));
}

export class BaselineManager {
  private _store: EvidenceStore;
  private _config: ResolvedConfig;
  private _detector: DriftDetector;
  private _tenantId: string | undefined;
  private _initialized: Promise<void> | null = null;

  constructor(store: EvidenceStore, config: ResolvedConfig, options?: { tenantId?: string }) {
    this._store = store;
    this._config = config;
    this._detector = new DriftDetector();
    this._tenantId = options?.tenantId;
  }

  private get agentId(): string {
    return (this._config.agentName as string | undefined) || "default";
  }

  private ensureTable(): Promise<void> {
    if (this._initialized) return this._initialized;
    this._initialized = (async () => {
      await this._store.exec(CREATE_BASELINES_TABLE_SQL);
      // Migrate: add tenant_id column if absent
      const cols = await this._store.query("PRAGMA table_info('baselines')");
      const hasCol = (cols as Array<Record<string, unknown>>).some(c => c.name === "tenant_id");
      if (!hasCol) {
        await this._store.exec("ALTER TABLE baselines ADD COLUMN tenant_id VARCHAR");
      }
      await this._store.exec("CREATE INDEX IF NOT EXISTS idx_baselines_agent_active ON baselines(agent_id, is_active)");
      await this._store.exec("CREATE INDEX IF NOT EXISTS idx_baselines_agent_overlay ON baselines(agent_id, overlay_id)");
    })();
    return this._initialized;
  }

  async create(options: {
    label: string;
    overlayId?: string;
    evidenceWindowHours?: number;
    metadata?: Record<string, unknown>;
  }): Promise<Baseline> {
    await this.ensureTable();
    const { label, overlayId, evidenceWindowHours = 168, metadata } = options;
    const agentId = this.agentId;
    const now = new Date();
    const windowStart = new Date(now.getTime() - evidenceWindowHours * 3600 * 1000).toISOString();
    const windowEnd = now.toISOString();

    const conditions: string[] = ["agent_id = ?", "timestamp >= ?"];
    const params: unknown[] = [agentId, windowStart];
    if (this._tenantId) {
      conditions.push("tenant_id = ?");
      params.push(this._tenantId);
    }
    if (overlayId !== undefined) {
      conditions.push("list_contains(CAST(active_overlays AS VARCHAR[]), ?)");
      params.push(overlayId);
    }

    const where = ` WHERE ${conditions.join(" AND ")}`;
    const rows = await this._store.query(
      `SELECT control_results FROM evidence_records${where}`,
      params,
    );

    const stats = computeControlStats(rows as Array<{ control_results: unknown }>);

    const snapshots: ControlSnapshot[] = Object.entries(stats).map(([cid, s]) => ({
      controlId: cid,
      result: dominantResult(s),
      passRate: passRate(s),
      totalEvaluations: s.total,
      evidenceWindowStart: windowStart,
      evidenceWindowEnd: windowEnd,
    }));

    // Deactivate existing active baseline for same agent+overlay
    if (overlayId !== undefined) {
      await this._store.run(
        "UPDATE baselines SET is_active = FALSE WHERE agent_id = ? AND overlay_id = ? AND is_active = TRUE",
        [agentId, overlayId],
      );
    } else {
      await this._store.run(
        "UPDATE baselines SET is_active = FALSE WHERE agent_id = ? AND overlay_id IS NULL AND is_active = TRUE",
        [agentId],
      );
    }

    const baselineId = randomUUID();
    const createdAt = now.toISOString();
    const baseline: Baseline = {
      baselineId,
      createdAt,
      agentId,
      overlayId: overlayId ?? null,
      label,
      controlSnapshots: snapshots,
      metadata: metadata ?? null,
      isActive: true,
    };

    await this._store.run(
      "INSERT INTO baselines (baseline_id, created_at, agent_id, overlay_id, label, control_snapshots, metadata, is_active, tenant_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
      [
        baseline.baselineId,
        baseline.createdAt,
        baseline.agentId,
        baseline.overlayId,
        baseline.label,
        snapshotsToJson(baseline.controlSnapshots),
        baseline.metadata !== null ? JSON.stringify(baseline.metadata) : null,
        baseline.isActive,
        this._tenantId ?? null,
      ],
    );

    return baseline;
  }

  async listBaselines(overlayId?: string): Promise<Baseline[]> {
    await this.ensureTable();
    const agentId = this.agentId;
    const conditions: string[] = ["agent_id = ?"];
    const params: unknown[] = [agentId];

    if (this._tenantId) {
      conditions.push("tenant_id = ?");
      params.push(this._tenantId);
    }
    if (overlayId !== undefined) {
      conditions.push("overlay_id = ?");
      params.push(overlayId);
    }

    const where = ` WHERE ${conditions.join(" AND ")}`;
    const rows = await this._store.query(
      `SELECT baseline_id, created_at, agent_id, overlay_id, label, control_snapshots, metadata, is_active FROM baselines${where} ORDER BY created_at DESC`,
      params,
    );
    return (rows as Array<Record<string, unknown>>).map(rowToBaseline);
  }

  async getBaseline(baselineId: string): Promise<Baseline> {
    await this.ensureTable();
    const rows = await this._store.query(
      "SELECT baseline_id, created_at, agent_id, overlay_id, label, control_snapshots, metadata, is_active FROM baselines WHERE baseline_id = ?",
      [baselineId],
    );
    const row = (rows as Array<Record<string, unknown>>)[0];
    if (!row) throw new Error(`Baseline not found: ${baselineId}`);
    return rowToBaseline(row);
  }

  async checkDrift(options?: { baselineId?: string; overlayId?: string }): Promise<DriftReport> {
    await this.ensureTable();
    const agentId = this.agentId;
    let baseline: Baseline;

    if (options?.baselineId) {
      baseline = await this.getBaseline(options.baselineId);
    } else {
      const conditions: string[] = ["agent_id = ?", "is_active = TRUE"];
      const params: unknown[] = [agentId];
      if (this._tenantId) {
        conditions.push("tenant_id = ?");
        params.push(this._tenantId);
      }
      if (options?.overlayId !== undefined) {
        conditions.push("overlay_id = ?");
        params.push(options.overlayId);
      }
      const where = ` WHERE ${conditions.join(" AND ")}`;
      const rows = await this._store.query(
        `SELECT baseline_id, created_at, agent_id, overlay_id, label, control_snapshots, metadata, is_active FROM baselines${where} ORDER BY created_at DESC LIMIT 1`,
        params,
      );
      const row = (rows as Array<Record<string, unknown>>)[0];
      if (!row) throw new Error("No active baseline found. Run 'ancilis baseline create' first.");
      baseline = rowToBaseline(row);
    }

    const since = baseline.createdAt;
    const conditions: string[] = ["agent_id = ?", "timestamp >= ?"];
    const params: unknown[] = [agentId, since];
    if (this._tenantId) {
      conditions.push("tenant_id = ?");
      params.push(this._tenantId);
    }
    if (baseline.overlayId !== null) {
      conditions.push("list_contains(CAST(active_overlays AS VARCHAR[]), ?)");
      params.push(baseline.overlayId);
    }
    const where = ` WHERE ${conditions.join(" AND ")}`;

    const evidenceRows = await this._store.query(
      `SELECT control_results, evaluation_id, tool_name, timestamp FROM evidence_records${where}`,
      params,
    );

    // No evaluations since baseline → posture unchanged, report STABLE immediately.
    const typedEvidenceRows = evidenceRows as Array<Record<string, unknown>>;
    if (typedEvidenceRows.length === 0) {
      const { randomUUID } = await import("node:crypto");
      return {
        driftReportId: randomUUID(),
        baselineId: baseline.baselineId,
        baselineLabel: baseline.label,
        checkedAt: new Date().toISOString(),
        agentId: baseline.agentId,
        overlayId: baseline.overlayId,
        overallStatus: "STABLE",
        summary: {
          totalControls: baseline.controlSnapshots.length,
          regressed: 0,
          degraded: 0,
          stable: baseline.controlSnapshots.length,
        },
        controlDrifts: [],
      };
    }

    const currentStats = computeControlStats(
      typedEvidenceRows.map(r => ({ control_results: r.control_results })),
    );

    // First failure timestamps per control — single-pass over already-fetched rows, no N+1
    const firstFailures: Record<string, string | null> = {};
    for (const cid of Object.keys(currentStats)) {
      firstFailures[cid] = null;
    }
    for (const row of typedEvidenceRows) {
      const crRaw = row.control_results;
      const crs = typeof crRaw === "string"
        ? JSON.parse(crRaw) as Array<Record<string, unknown>>
        : crRaw as Array<Record<string, unknown>>;
      for (const cr of crs) {
        const cid = cr.control_id as string;
        const result = ((cr.result as string | undefined) ?? "SKIP").toUpperCase();
        if (result === "FAIL" || result === "ERROR") {
          const ts = row.timestamp as string;
          if (firstFailures[cid] === null || ts < firstFailures[cid]!) {
            firstFailures[cid] = ts;
          }
        }
      }
    }

    // New failure evaluation IDs and tools per control
    const newFailureIds: Record<string, string[]> = {};
    const failureTools: Record<string, string[]> = {};
    for (const row of typedEvidenceRows) {
      const crRaw = row.control_results;
      const crs = typeof crRaw === "string"
        ? JSON.parse(crRaw) as Array<Record<string, unknown>>
        : crRaw as Array<Record<string, unknown>>;
      for (const cr of crs) {
        const cid = cr.control_id as string;
        const result = ((cr.result as string | undefined) ?? "SKIP").toUpperCase();
        if (result === "FAIL" || result === "ERROR") {
          if (!newFailureIds[cid]) newFailureIds[cid] = [];
          if (!failureTools[cid]) failureTools[cid] = [];
          newFailureIds[cid]!.push(row.evaluation_id as string);
          const toolName = row.tool_name as string;
          if (!failureTools[cid]!.includes(toolName)) {
            failureTools[cid]!.push(toolName);
          }
        }
      }
    }

    return this._detector.detect(baseline, currentStats, firstFailures, newFailureIds, failureTools);
  }

  async deactivate(baselineId: string): Promise<void> {
    await this.ensureTable();
    await this._store.run(
      "UPDATE baselines SET is_active = FALSE WHERE baseline_id = ?",
      [baselineId],
    );
  }
}
